Edit/create shadows in global styles

madhusudhand opened the pull request:
## What?

This change introduces a new item `shadows` in the global styles for **editing/creating shadows**

<img width="285" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/1c9a6a9a-0e13-4f95-860e-87618df72ed3">
<img width="286" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/0381da3e-60b4-4758-9f6f-a6e26c7482e6">
<img width="841" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/c002f1c2-1b39-4188-8fc6-432cd2e9fcee">

## Testing instructions

1. Find newly added item `shadows` in global styles panel
2. Click to land on shadows panel, it should show list of `core`, `theme` and `custom` shadows.
3. If there are not `theme` or `custom` shadows, it should show a placeholder label.
4. Select a shadow to land on preview screen

Editing a shadow:

1. In the preview screen, click on `Drop shadow` to invoke editor, change lengths, color, inset etc.. and preview should update instantly.
2. Click `+` to add new shadow layer.
3. click `-` to remove a shadow layer. (this is available when there are at least 2 layers)
4. Save the edits.

## Subtasks

- [x] Show list of `default`, `theme` and `custom` shadows.
- [x] Preview shadow
- [x] Update a shadow (editor panel)
- [x] Add a shadow
- [x] Delete a shadow
- [ ] Duplicate a shadow
- [x] Rename a shadow (how about slug??, also handle conflicts)
- [x] add or delete a shadow layer (in second screen)

## Related issues:

fixes: #57100

madhusudhand committed:
add shadows panel in global styles

madhusudhand committed:
add shadows editor view

madhusudhand committed:
add color edit popover

madhusudhand committed:
refine ui between default and custom shadows

madhusudhand committed:
update shadows UI to match new designs

madhusudhand committed:
add unit tests

aaronrobertshaw committed:
Try different approach to parsing shadow strings

aaronrobertshaw commented:
@madhusudhand as promised I've pushed a different approach to parsing shadow strings in https://github.com/WordPress/gutenberg/pull/60706/commits/60c40f3e5c5536577dedba7c0ca508d6d51d6317.

While that commit gets all 2040 unit tests passing, it was only quickly hacked together and could definitely do with some refinement. That said, hopefully, it helps unblock the rest of the UI work here, I'm looking forward to seeing the end result!

Some aspects of the updated approach that need looking at are;
- making sure that there is type consistency in parsed values e.g. `'0'` vs `0`
- making it more robust in the face of invalid strings, including extra unit tests for these
- inline comments to explain the various regex patterns used and maybe also the fact that length values must all be together for it to be a valid shadow string
- whether or not extra spaces should be stripped when the shadow strings are parsed or built
- performance, there's likely a more performant approach that this could be refined towards

There's probably more that I haven't noted but as this PR is still in its early days, I'm sure it will all come out in the wash 😄

madhusudhand committed:
add more unit tests and improve util functions

madhusudhand committed:
update shadows edit panel

madhusudhand committed:
fix unit conversion issues and other ui improvements

madhusudhand commented:
## Short summary for design review

Changes are now ready for the first design review. While it addresses everything as per the suggested designs, following are done differently.

### Popover

| Design | Actual UI |
| -- | -- |
| <img width="259" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/2fa9a418-4545-4c74-9c24-0a12a9b0ae77"> | <img width="291" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/ada23416-1c6f-44c7-9047-2a4b76bde470"> |

1. For the color selection `ColorPalette` core component is used and it comes the the above layout, and differs from the suggested one.
2. `x,y position square` is not implemented yet in the first iteration. as it requires a new component and needs to be tested for acceptability, it shall be implemented in the next iteration.

❓ What do you suggest as the boundaries for the slider. 
In the current revision `0-20px` or `0-10rem` or `0-100%` are used as boundaries.

### Preview

| Design | Actual UI |
| -- | -- |
| <img width="290" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/d36a36e4-c5f5-4384-a438-d7fc725be326"> | <img width="284" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/3ccdeb11-fbeb-4213-8f57-6aec15dbf280"> |

A white background is used by taking some inspiration from the block preview component. 
❓Please confirm if should be a transparent background.

### Shadow options

<img width="287" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/39de1ec7-0ede-46ae-824c-5289a32e2ed2">

❓Could you suggest designs for the `rename`, `duplicate` options. And for `delete` do you recommend having a confirmation UI? If yes, could you provide one?

### Shadow item labeling

<img width="283" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/72a72681-b247-4b9a-bad9-8870e37967a9">

❓When the inset is true, do you think the above labeling makes sense?

### Shadows panel copy

<img width="286" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/90c4b6c4-b8e5-4062-b14d-cbfaa8824746">

Please provide copy.

jarekmorawski commented:
> x,y position square is not implemented yet in the first iteration. as it requires a new component and needs to be tested for acceptability, it shall be implemented in the next iteration.

Would it make sense to use the [AlignmentMatrixControl](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-alignmentmatrixcontrol--docs) component before we build a proper one?

jameskoster commented:
Thanks for the ping @madhusudhand, excellent work here, this is coming together nicely :)

> For the color selection ColorPalette core component is used and it comes the the above layout, and differs from the suggested one.

Good idea, makes sense.

> x,y position square is not implemented yet in the first iteration. as it requires a new component and needs to be tested for acceptability, it shall be implemented in the next iteration.

Also fine, and no rush on this.

> What do you suggest as the boundaries for the slider.

0 - 20px seems a reasonable range to begin with – it's easy to tweak this later. I wonder if it's worth capturing negative values? Maybe we try `-20` - `20` with `0` as the default?

> A white background is used by taking some inspiration from the block preview component.
>❓Please confirm if should be a transparent background.

I think it's worth trying the transparent texture so that white shadows are somewhat visible.

> ❓Could you suggest designs for the rename, duplicate options. And for delete do you recommend having a confirmation UI? If yes, could you provide one?

Could we use [DropdownMenu V2](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2--docs) for this menu?

Rename should open a `small` [Modal](https://wordpress.github.io/gutenberg/?path=/docs/components-modal--docs), same as renaming a page, template, or pattern:

<img width="371" alt="Screenshot 2024-05-02 at 15 22 06" src="https://github.com/WordPress/gutenberg/assets/846565/53bf6d6a-23d5-4baa-97c8-3099a691040b">

Delete should open a [ConfirmDialog](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-confirmdialog--docs), same as deleting a template:

<img width="440" alt="Screenshot 2024-05-02 at 15 24 15" src="https://github.com/WordPress/gutenberg/assets/846565/1a947180-b18e-480c-ae76-dbbffc5ae26d">

**Important**: Only user-generated (custom) shadows should be rename-able or delete-able.

Duplicate is a little tricky because we need to think about where to send the user post duplication. I'd be tempted to leave duplication out of this PR, and revisit it separately.

> ❓When the inset is true, do you think the above labeling makes sense?

Yup, looks good :) 

> Please provide copy.

Maybe something like: "Manage and create shadow styles for use across the site.".

---
Some additional points:

* When a theme or default shadow has been edited we probably need to add a 'Reset' action to the menu.
* The `–` icon for removing shadows should be a Button and include a tooltip.
* When there are no theme shadows we can hide the "Theme" section
* I don't think we need to show the "Custom shadows are empty!" message.

richtabor commented:
@jasmussen Here's another proposed top-level navigation item for global styles. 

I think we should map out how presets like shadow (and spacing eventually) should be presented in global styles.

jasmussen commented:
Agreed, the current IA is not scaling well. The issue feels a bit pressed by the too-granular "Background image". Going back to #47369, the main instinct there is to separate in two categories, presets and styles. Saxon suggests this:

> Presets
> 
> * Typography
> * Colors
> * Spacing
> * Layout
> 
> Styles
> 
> * Elements
> * Blocks

"Shadows" is perhaps not the best term, but since it's a set of presets, the above heuristic would provide a structure for it at the top, alongside Colors and Typography.

Layout is a bit tricky since it's in-between presets (through inheritance of block spacing and content widths), and styles for the current page (padding). I wonder if we could split that out, clean it up once and for all. The following is a quick idea intended to inspire better ideas:

> Presets
> 
> * Typography
> * Colors
> * Shadows
> * Spacing
> 
> Styles
> 
> * Site
> * Elements
> * Blocks

"Background image" would go in "Site". As would "Padding", currently sitting in Layout. Block spacing would move from Layout to Spacing, along with other presets eventually. "Site" could also be a place where _Border_ could be added.

jameskoster commented:
The separation makes sense, but unsure about "Styles" and "Presets" as headings. Would a shadow be a style or a preset... it's kind of both? The headings might not both be necessary, the labels are generally self-explanatory. 

Either way, probably a discussion for https://github.com/WordPress/gutenberg/issues/53483. Saxon shared some interesting ideas there already.

jameskoster commented:
I guess it tripped me up because shadows can be both a style, and a style preset. Maybe I'm too used to the 'styles' concept in Figma.

"Background" and "Layout" being grouped together under "Site" seems worth a try.

jorgefilipecosta commented:
Hi @madhusudhand,
It seems that although custom shadows are stored:
<img width="279" alt="Screenshot 2024-05-03 at 18 14 39" src="https://github.com/WordPress/gutenberg/assets/11271197/689f3be7-40bb-45a4-ae16-467ed37a915a">

They are not shown as one of the available options to pick:
<img width="279" alt="Screenshot 2024-05-03 at 18 14 39" src="https://github.com/WordPress/gutenberg/assets/11271197/9ca3b0cf-ddd0-4f9d-a45e-fab8a7577828">



For colors, we show the multiple origins of colors:
<img width="242" alt="Screenshot 2024-05-03 at 18 19 47" src="https://github.com/WordPress/gutenberg/assets/11271197/ef5c1bc4-27a7-43fd-ac43-c96622dcaf09">
Should we do the same for shadows?

Are there any plans to allow to pick a specific shadow in a given place besides the preset ones?

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/shadows-edit-panel.js:
```diff
@@ -0,0 +1,441 @@
+		offset: 36,
+		shift: true,
+	};
+	const [ shadowObj, setShadowObj ] = useState(
```
Nitpick: Using the local state may be problematic. For example, a plugin may update the pallet object directly, and our local state will keep the old shadow object. For collaborative editing, this will also create a bug, if two users have the same shadow opened, changes one user makes will not propagate because although the entity is changed the local state will not be changed.

It may be better to just remove the local state and a useMemo that computes the object using shadowStringToObject. Making the style object the only source of truth.

madhusudhand replied:
Great catch. Thank you. 
Removed the local state here 719ca407ec889c5a871fde596fa7401e33e7188d

jorgefilipecosta commented:
Not for this PR but, Would it make sense when picking a color for the shadow show the preset color palettes? If a preset color is selected the shadow would reference it by its variable making the shadows dynamically change when the colors change.

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/shadows-panel.js:
```diff
@@ -0,0 +1,139 @@
+	return (
+		<NavigationButtonAsItem
+			path={ `/shadows/edit/${ category }/${ shadow.slug }` }
+			aria-label={ 'edit' }
```
Could this aria-label be more descriptive and specify something like "Edit shadow YYY". It should also be translatable using "__".

madhusudhand replied:
Fixed here 719ca407ec889c5a871fde596fa7401e33e7188d

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/shadows-panel.js:
```diff
@@ -0,0 +1,139 @@
+function ShadowList( { label, placeholder, shadows, category, onCreate } ) {
+	const handleAddShadow = () => {
+		onCreate( {
+			name: `Shadow ${ shadows.length + 1 }`,
```
Probably "Shadow" should be translatable.

madhusudhand replied:
It is a name given by user such as `My custom shadow`, and when creating a first shadow giving it a default name `Shadow n`. So, it is not a candidate for translations.

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/shadows-panel.js:
```diff
@@ -0,0 +1,139 @@
+						shadows={ defaultShadows || [] }
+						category={ 'default' }
+					/>
+					<ShadowList
```
Not a blocker but I think for each of the shadow groups "Default" and "Theme" we should allow the user to reset the changes as we do for colors.

madhusudhand replied:
Addressed in 6e3833951af89e752138b8f8c21efad5735f807e

jorgefilipecosta commented on packages/edit-site/src/components/global-styles/shadows-edit-panel.js:
```diff
@@ -0,0 +1,441 @@
+];
+
+export default function ShadowsEditPanel() {
+	const {
```
Not, a blocker as the actions are not yet implemented but things like Rename, Duplicate, and Delete should only appear on the custom "group" right now they also appear on default and theme.

madhusudhand replied:
Addressed in 3defb6a6e4507fe7246bc01206262175ff4a7edf

jorgefilipecosta reviewed (commented):
Hi @madhusudhand, the main functionality is working well, and the code looks good 👍  Any specific part where a deeper look may be helpful?

madhusudhand committed:
add shadow rename and delete functions

madhusudhand committed:
address design feedback

madhusudhand committed:
add option to reset default and theme shadows

madhusudhand commented:
@jameskoster Thank you for the feedback. Following are the changes.

> 0 - 20px seems a reasonable range to begin with – it's easy to tweak this later. I wonder if it's worth capturing negative values? Maybe we try -20 - 20 with 0 as the default?

`X`, `Y` and `spread` now supports negative values.

> I think it's worth trying the transparent texture so that white shadows are somewhat visible.

<img width="273" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/7ad267c9-86fd-4950-a8e3-0e5db9f5bc1e">

> Could we use [DropdownMenu V2](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2--docs) for this menu?

Updated to V2 dropdown menu

> Rename should open a small [Modal](https://wordpress.github.io/gutenberg/?path=/docs/components-modal--docs), same as renaming a page, template, or pattern:

<img width="402" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/827b5339-772d-4bb2-87b4-cc28d086f5c8">

> Delete should open a [ConfirmDialog](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-confirmdialog--docs), same as deleting a template:

<img width="423" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/46a1ba01-6b98-4570-b11e-6874288335d0">

> Important: Only user-generated (custom) shadows should be rename-able or delete-able.

Yes, this is taken care. Only custom shadows will have menu options.

> When a theme or default shadow has been edited we probably need to add a 'Reset' action to the menu.

<img width="285" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/d07d3b34-e3a5-45f2-9ba7-21316842e6f6">

> When there are no theme shadows we can hide the "Theme" section
> I don't think we need to show the "Custom shadows are empty!" message.

Addressed in 318b70ace276619e03c8b67b21b5ad87e4bc08a5

> The – icon for removing shadows should be a Button and include a tooltip.

<img width="292" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/3419e0bb-392f-487d-8273-8c70144aeaa3">

The wrapping component is a button, and having a button inside a button brings focus, but the tooltip of the inner button doesn't work. 
❓ Do you suggest to split them into two buttons in the same level?

madhusudhand committed:
add custom shadows in block styles

madhusudhand committed:
remove local state and use momoize

madhusudhand committed:
fix lint issue

madhusudhand commented:
@jorgefilipecosta Thank you for the review. Addressed the review findings and could you take another look.

> It seems that although custom shadows are stored:, They are not shown as one of the available options to pick:

Fixed here df0238da006b65f7f5352dfe253fd6252b89c1e6

> There is also a small issue where if the shadow is too big it may make it impossible to correctly preview the other shadows.

The shadow was restricted with in a boundary in the first iteration. However it was later changed in #58663 to overflow.

> For colors, we show the multiple origins of colors: Should we do the same for shadows?

Created #61441 for discussion.

> the main functionality is working well, and the code looks good 👍 Any specific part where a deeper look may be helpful?

[Current implementation](https://github.com/WordPress/gutenberg/pull/60706/files#diff-d76fe0b9848e786a8cc26155bd54635fb31ed7bb5ba193af1e702fe7c1cbeaaeR132) uses `shadow-${ shadows.length + 1 }` as slug. But that would be a conflict when there exists same slug in `theme.json` (also slug is used as key in react map in block styles, it will through errors)
Do you have any thoughts on avoiding slug conflicts?

jasmussen commented:
Nice work, this is close indeed.

A few quick notes, apologies if there's overlap with what has already been noted. Coming to mind yet again, we do need to make progress on https://github.com/WordPress/gutenberg/pull/60706#issuecomment-2092527575. Should I open a separate issue or can we follow up? A plan B is to put it all inside "Layout" along with "Background image", if we don't get to add the presets, etc. 

On the ItemGroup items, I'm seeing a small double border at the bottom:

![shadows](https://github.com/WordPress/gutenberg/assets/1204802/316f8e5b-ce23-4b94-bebd-7fb3905ab9d5)

It's not clear to me that this is the result of this branch or whether it's present in trunk or not. But if you click the item, drag outside, then the double border shows up. Mainly important to address if it's a result of this branch.

These items _look_ like ItemGroups, but it's something else:

![Screenshot 2024-05-08 at 08 29 14](https://github.com/WordPress/gutenberg/assets/1204802/0ec9eb6c-0863-4359-b236-df23e810633c)

This is okay as a temporary thing if it gets us shipping. But to be clear, the [ItemGroup](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-itemgroup--docs) itself needs to support the "minus" functionality. This is something @SantosGuillamot has been looking at, as well, for a "block connections" panel. So if there's an opportunity to sync up there, or move this to using an ItemGroup in the future, we should at least track this. 

If we go with the custom temporary component for now, it'd be nice to get the _height_ closer. ItemGroup items are supposed to be exactly 40px tall (like the default button size), but there's a bug in them at the moment meaning they're slightly taller:

![Screenshot 2024-05-08 at 08 29 36](https://github.com/WordPress/gutenberg/assets/1204802/33ec3af4-0fd7-4f8f-b863-0974db2dfcad)

But this one is 45px tall instead of 41.8px tall:

![Screenshot 2024-05-08 at 08 29 22](https://github.com/WordPress/gutenberg/assets/1204802/f60dbba4-3436-49e8-ae43-b43581e9c000)

All of this can potentially be followup territory. Just important we track it.

jameskoster commented:
> Should I open a separate issue or can we follow up? 

Imo we should work on the iA parallel to this work. This PR doesn't seem like a great spot to concentrate on that, especially as it involves moving the Background image property which is totally separate. I wouldn't expect to find shadow presets inside "Layout", personally.

jasmussen commented:
> Imo we should figure out the iA parallel to this work. This PR doesn't seem like a great spot to concentrate on that, especially as it involves moving the Background image property which is totally separate. I wouldn't expect to find shadow presets inside "Layout", personally.

It's definitely separate. It's just someting we need figured out for 6.6. I'd also agree Layout isn't the best place for "Background image", but it's better than a revert, which would be a 6.6 _plan B_. Do you think Saxon's IA issue is the right place, or do we need a smaller 6.6-bound interim step? Sorry about the detour! And nice work, all!

jameskoster commented:
> or do we need a smaller 6.6-bound interim step?

I opened https://github.com/WordPress/gutenberg/issues/61494 to capture this specifically 👍

madhusudhand committed:
move reset option from shadows panel to shadow edit panel

madhusudhand committed:
split shadow dropdown button into two buttons

madhusudhand commented:
@jameskoster 

> Apologies, I meant the menu in the edit-shadow panel, this one:
Basically the same ellipsis button as here, but with a "Reset" action instead of "Rename" and "Delete":

| Preset shadow options | Custom shadow options |
| -- | -- |
| <img width="285" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/2367e781-42d1-4fa7-9ca4-86354bef7841"> | <img width="285" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/02ac0637-fe66-4082-b730-c2ee98c772e0"> |

Updated the menu options. `Reset` is enabled only when there are modifications to `defautl` or `theme` shadow.

> Thanks for adding the rename action, I wonder if we should validate to avoid duplicate names? We don't necessarily have to do it in this PR, but if it's trivial to implement then one option would be to disable the Save button if the text control value matches an existing name onChange. What do you think?

I think its a good idea. Also, instead of just disabling, the dialog can show an error message as to why the `name` is not allowed. Created #61620 to followup on this.

> Good point, we should avoid nesting interactive elements. We essentially want to create something like this, do you think it's possible?

Yes, I have split it into two buttons and following is the UI.

![shadow-split-button](https://github.com/WordPress/gutenberg/assets/1935113/b532e032-a1df-40f7-bd40-39df52173398)

> A tiny point—it looks like the ellipsis buttons are using the tertiary variant. Ideally these use the default variant to match other icon buttons.

Thanks for pointing. Updated to use default variant.


@jasmussen 
> On the ItemGroup items, I'm seeing a small double border at the bottom:

I tested the issue in other places. It seem to be issue from the core component `NavigationButtion` when inside ItemGroup.
Opened #61621

<img width="289" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/78d2cc00-ab3b-4f43-a678-5e2945821ec5">

> This is okay as a temporary thing if it gets us shipping. But to be clear, the [ItemGroup](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-itemgroup--docs) itself needs to support the "minus" functionality. This is something @SantosGuillamot has been looking at, as well, for a "block connections" panel. So if there's an opportunity to sync up there, or move this to using an ItemGroup in the future, we should at least track this.

I support this idea, and sure I will syncup with @SantosGuillamot on this.

> But this one is 45px tall instead of 41.8px tall:

<img width="309" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/49fb8d6e-5466-4080-97b1-df5abe59dd31">

jameskoster commented:
Thanks for the update!

> I support this idea

I wonder if we should update ItemGroup to support the two-button configuration first, then utilise that here? Otherwise we're going to need to add some style overrides;

* Set the height to 40px
* Fix the minus button dimensions to 40x40
* Add an active state to the row when the popover is open, to match other UIs e.g. color:

<img width="570" alt="Screenshot 2024-05-13 at 13 50 57" src="https://github.com/WordPress/gutenberg/assets/846565/752b85be-eb14-4f8c-9859-d4710bbef589">

---

In the shadow configuration popover could the segmented control / range / unit picker controls use the `__next40pxDefaultSize` property?

---

Could the minus button use the `reset` icon?

---

When deleting a shadow:

> Are you sure you want to delete ~the shadow~ "Shadow 3"?

---

Is there some logic about the properties of the shadow when adding new shadows to a style, or is it always the same? If it's always the same, we should probably use a semi-transparent color value. This will make it a bit clearer in the preview that the shadow was added successfully.

madhusudhand committed:
fix item height to 40px

madhusudhand committed:
validate invalid shadow strings

madhusudhand commented on packages/edit-site/src/components/global-styles/shadow-utils.js:
```diff
@@ -0,0 +1,158 @@
+	 * 3. Should not contain more than one set of x, y, blur, spread values.
+	 * 4. Should contain at least x and y values. Others are optional.
+	 * 5. Should not contain more than one "inset" (case insensitive) keyword.
+	 * 6. Should not contain more than one color value.
```
@aaronrobertshaw Updated the function with detailed notes and also, the same rules are kept inline at every step of the function. Should we add/update anything here?

Also, updated the function definition to handle invalid shadow formats. Let me know if it can be improved further.

madhusudhand committed:
update spacing

madhusudhand commented:
@jameskoster Thanks a lot for the excellent reviews, and sorry that minor issues are missed and it has to go through multiple iterations.

> There's a small inconsistency between headings and item groups as illustrated in the following screenshot:

Updated the gap to constant `8px` all over.

> The UI jumps slightly when toggling between range/text input:

Removed bottom margin for the range control. However, the controls in the last row causes slight overflow because its tooltip is going out of bounds.
Had to give additional margin of `4px` for the container to counter that.

> I'm seeing an empty label span above the segmented control:

Fixed.

jasmussen commented:
> Seems to be because the range has a small bottom margin while the text input does not. The gap between grid columns is also a little large. I suggest removing the margin on the range, and setting the grid spacing to 16px (on both axis).

We need to fix this on the componentry level. Essentially the Range should default to a 40px footprint, exactly so it can sit next to a 40px input like here. And it could/should then have a `compact` prop that makes it 32px tall.

May not have to be fixed here, but that's also the answer to spacing: things should be spaced out as if the range unit was 40px tall.

jameskoster reviewed (approved):
This looks ready from a design perspective. There may be small tweaks to make, but they can be polished in follow-ups. It'd be good to get this in sooner rather than later with the beta on the horizon.

Nice work @madhusudhand!

mikachan commented on packages/edit-site/src/components/global-styles/test/shadow-utils.spec.js:
```diff
@@ -0,0 +1,254 @@
+	hslp: 'hsl(0 100% 50% / 50%)',
+	currentcolor: 'currentcolor',
+
+	// invalidColor1: 'invalid',
+	// invalidColor2: 'hwb(0, 0%, 0%)',
```
I noticed these comments have been left in and are not mentioned elsewhere - should they be removed?

madhusudhand replied:
These are unnecessary in the present context. removed them.

mikachan reviewed (approved):
I've followed the testing instructions and this PR is working really well. Great work, @madhusudhand! 🎉

- Can see the newly added item "Shadows" in the global styles panel ✅
- Can successfully navigate to the shadow preview panel ✅
- Can edit a shadow and change lengths, colour, inset, and the shadow preview updates instantly ✅
- Can add and remove shadows ✅
- Can rename a shadow ✅

madhusudhand committed:
delete comments

madhusudhand commented:
Thanks for the excellent reviews. Bringing this feature in and more suggestions/issues will be addressed in followup PRs.

madhusudhand commented:
I am sorry, I missed copying props.
